controllers/storageclassclaim/storageclassclaim_controller.go

```diff
@@ -92,20 +137,319 @@ func (r *StorageClassClaimReconciler) Reconcile(ctx context.Context, request rec
+			},
+			Finalizers: []string{
+				StorageClassFinalizer,
+			},
+		},
+		ReclaimPolicy:        &pvReclaimPolicy,
+		AllowVolumeExpansion: &allowVolumeExpansion,
+		Provisioner:          fmt.Sprintf("%s.cephfs.csi.ceph.com", r.storageCluster.Namespace),
+		Parameters:           data,
+	}
+	return storageClass
+}
+
+func (r *StorageClassClaimReconciler) getCephRBDStorageClass(data map[string]string) *storagev1.StorageClass {
+	pvReclaimPolicy := corev1.PersistentVolumeReclaimDelete
+	allowVolumeExpansion := true
+	storageClass := &storagev1.StorageClass{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      r.storageClassClaim.Name,
+			Namespace: r.storageClassClaim.Namespace,
+			Annotations: map[string]string{
+				"description": "Provides RWO Filesystem volumes, and RWO and RWX Block volumes",
+			},
+			Finalizers: []string{
+				StorageClassFinalizer,
+			},
+		},
+		ReclaimPolicy:        &pvReclaimPolicy,
+		AllowVolumeExpansion: &allowVolumeExpansion,
+		Provisioner:          fmt.Sprintf("%s.rbd.csi.ceph.com", r.storageCluster.Namespace),
+		Parameters:           data,
+	}
+	return storageClass
+}
+
+func (r *StorageClassClaimReconciler) createOrReplaceStorageClass(storageClass *storagev1.StorageClass) error {
+	existing := &storagev1.StorageClass{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: r.storageClassClaim.Name,
+		},
+	}
+
+	// Check if the StorageClass already exists, if not create a new one.
+	if err := r.get(existing); errors.IsNotFound(err) {
+		r.log.Info("Creating StorageClass.", "StorageClass", klog.KRef(storageClass.Namespace, existing.Name))
+		err := r.Client.Create(r.ctx, storageClass)
+		if err != nil {
+			return fmt.Errorf("failed to create StorageClass: %v", err)
+		}
+	} else if err != nil {
+		return fmt.Errorf("failed to get StorageClass: %v", err)
+	}
+
+	// If present then compare the existing StorageClass with the received StorageClass, and only proceed if they differ.
+	if reflect.DeepEqual(existing.Parameters, storageClass.Parameters) {
+		return nil
+	}
+
+	// StorageClass already exists, but parameters have changed. Delete the existing StorageClass and create a new one.
+	if existing.UID != "" {
+
+		// Since we have to update the existing StorageClass, so we will delete the existing storageclass and create a new one.
+		r.log.Info("StorageClass needs to be updated, deleting it.", "StorageClass", klog.KRef(storageClass.Namespace, existing.Name))
+
+		// Check if StorageClass is in use, if yes, then fail.
+		// Wait until all PVs using the StorageClass under deletion are removed.
+		// Check for any PVs using the StorageClass.
+		pvList := corev1.PersistentVolumeList{}
+		err := r.list(&pvList, client.MatchingFields{
+			"spec.storageClassName": r.storageClassClaim.Name,
+		})
+		if err != nil {
+			return fmt.Errorf("failed to list PersistentVolumes: %s", err)
+		}
+		if len(pvList.Items) > 0 {
+			return fmt.Errorf("StorageClass %s is still in use by one or more PV(s)", r.storageClassClaim.Name)
+		}
```

Comment:
@rexagod I believe that after our discussion yesterday this can be removed as we got to the conclusion that deleting a storageclass will have no effect on existing PVCs/PVs.

 @Madhu-1 Can you please confirm?
 
 ```suggestion
```

Reply:
Can we keep this for now? Madhu seems to be OOO till 4th.

Reply:
no, as it removes the ability to update a storageclass when there are existing VPC/VP for no reason. 
We already discussed that with @Madhu-1 offline, I have asked him to confirm for the record, but we can skip his confirmation.